src/crypto/generators.cpp

```diff
@@ -0,0 +1,192 @@
+    ge_p2 temp_p2;
+    ge_p1p1 temp_p1p1;
+
+    cn_fast_hash(reinterpret_cast<const unsigned char*>(&x), sizeof(hash), h);
+    ge_fromfe_frombytes_vartime(&temp_p2, reinterpret_cast<const unsigned char*>(&h));
+    ge_mul8(&temp_p1p1, &temp_p2);
+    ge_p1p1_to_p3(&temp_p3, &temp_p1p1);
+    ge_p3_tobytes(to_bytes(point_out), &temp_p3);
+}
+//-------------------------------------------------------------------------------------------------------------------
+//-------------------------------------------------------------------------------------------------------------------
+static public_key reproduce_generator_G()
+{
+    // G = {x, 4/5 mod q}
+    fe four, five, inv_five, y;
+    fe_0(four);
+    fe_0(five);
+    four[0] = 4;
+    five[0] = 5;
+    fe_invert(inv_five, five);
+    fe_mul(y, four, inv_five);
+
+    public_key reproduced_G;
+    fe_tobytes(to_bytes(reproduced_G), y);
+
+    return reproduced_G;
+}
+//-------------------------------------------------------------------------------------------------------------------
+//-------------------------------------------------------------------------------------------------------------------
+static public_key reproduce_generator_H()
+{
+    // H = 8*to_point(keccak(G))
+    // note: this does not use the point_from_bytes() function found in H_p(), instead directly interpreting the
+    //       input bytes as a compressed point (this can fail, so should not be used generically)
+    // note2: to_point(keccak(G)) is known to succeed for the canonical value of G (it will fail 7/8ths of the time
+    //        normally)
+    ge_p3 temp_p3;
+    ge_p2 temp_p2;
+    ge_p1p1 temp_p1p1;
+
+    hash H_temp_hash{cn_fast_hash(to_bytes(G), sizeof(ec_point))};
+    (void)H_temp_hash;  //suppress unused warning
+    assert(ge_frombytes_vartime(&temp_p3, reinterpret_cast<const unsigned char*>(&H_temp_hash)) == 0);
+    ge_p3_to_p2(&temp_p2, &temp_p3);
+    ge_mul8(&temp_p1p1, &temp_p2);
+    ge_p1p1_to_p3(&temp_p3, &temp_p1p1);
+
+    public_key reproduced_H;
+    ge_p3_tobytes(to_bytes(reproduced_H), &temp_p3);
+
+    return reproduced_H;
+}
+//-------------------------------------------------------------------------------------------------------------------
+// Make generators, but only once
+//-------------------------------------------------------------------------------------------------------------------
+static void init_gens()
+{
+    std::call_once(init_gens_once_flag,
+        [&](){
+
+        // copy uint generator representations to standard 'ec_point' representations (char arrays)
+        memcpy(G.data, G_uint.bytes, 32);
+        memcpy(H.data, H_uint.bytes, 32);
+
+        // build ge_p3 representations of generators
+        const int G_deserialize = ge_frombytes_vartime(&G_p3, to_bytes(G));
+        const int H_deserialize = ge_frombytes_vartime(&H_p3, to_bytes(H));
+
+        (void)G_deserialize; assert(G_deserialize == 0);
+        (void)H_deserialize; assert(H_deserialize == 0);
```

Comment:
This is OK I guess since it's totally constant but this type of asserts means they only ever get caught if you build without -DNDEBUG (which I pretty much never do, but I'm not sure it's common). So it looks tested, but it might not in practice. So OK in this particular case, but something more dynamic would probably not be.